File: src/RM.Base/Logger.cs
using System.Runtime.CompilerServices;

namespace RM.Base;

public static class Logger {
  #region Properties -----------------------------------------------------------
  /// <summary>Set until which LogLevel should be logged</summary>
  public static LogLevel LogLevel { get; set; }

  public static bool LogToConsole { get; set; }
  public static bool LogToFile { get; set; } = true;

  public static Queue<LogMessage> LogMessages { get; } = [];
  #endregion

  #region Events ---------------------------------------------------------------
  public static event Action? OnLogMessageAdded;
  #endregion

  #region Variables ------------------------------------------------------------
  private static readonly string LogFilePath = Path.Combine(
    Common.GetAppDataPath(),
    "RepoManager.log"
  );
  #endregion

  #region Interface ------------------------------------------------------------
  public static void Info(string message, [CallerMemberName] string caller = "", [CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int lineNumber = 0)
    => Log(LogLevel.Info, message, caller, sourceFilePath, lineNumber);
  public static void Error(string message, [CallerMemberName] string caller = "", [CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int lineNumber = 0)
    => Log(LogLevel.Error, message, caller, sourceFilePath, lineNumber);
  public static void Error(string message, Exception exception, [CallerMemberName] string caller = "", [CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int lineNumber = 0)
    => Log(LogLevel.Error, $"{message} - {exception.Message}", caller, sourceFilePath, lineNumber);
  public static void Debug(string message, [CallerMemberName] string caller = "", [CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int lineNumber = 0)
    => Log(LogLevel.Debug, message, caller, sourceFilePath, lineNumber);
  public static void DebugGit(string message, [CallerMemberName] string caller = "", [CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int lineNumber = 0)
    => Log(LogLevel.DebugGit, message, caller, sourceFilePath, lineNumber);

  public static void OpenLogFile() {
    if (!LogToFile) return;
    var logDir = Path.GetDirectoryName(LogFilePath)!;
    if (!Directory.Exists(logDir))
      Directory.CreateDirectory(logDir);
    System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo {
      FileName = LogFilePath, UseShellExecute = true
    });
  }
  #endregion

  #region Implementation -------------------------------------------------------
  private static void Log(LogLevel level, string message, string caller, string source, int line) {
    if (level > LogLevel) return;
    var logMessage = new LogMessage(level, message, caller, Path.GetFileName(source), line);
    LogMessages.Enqueue(logMessage);
    if (LogToFile)
      InternalLogToFile(logMessage);
    if (LogToConsole)
      logMessage.ToConsole();
    OnLogMessageAdded?.Invoke();
  }

  private static void InternalLogToFile(LogMessage message) {
    var logDir = Path.GetDirectoryName(LogFilePath)!;
    if (!Directory.Exists(logDir))
      Directory.CreateDirectory(logDir);

    const long maxSize = 10 * 1024 * 1024; // 10MB
    if (File.Exists(LogFilePath) && new FileInfo(LogFilePath).Length > maxSize) {
      var backupPath = Path.Combine(logDir, "RepoManager.log.1");
      if (File.Exists(backupPath))
        File.Delete(backupPath);
      File.Move(LogFilePath, backupPath);
      File.Create(LogFilePath).Dispose();
    }
    File.AppendAllText(LogFilePath, message + Environment.NewLine);
  }
  #endregion
}

public class LogMessage(LogLevel level, string message, string caller, string source, int line) {
  private LogLevel Level { get; } = level;
  private string Message { get; } = message;
  private DateTime Timestamp { get; } = DateTime.Now;
  private string Caller { get; } = caller;
  private string Source { get; } = source;
  private int Line { get; } = line;

  public override string ToString() {
    if (Level == LogLevel.Error || Level >= LogLevel.Debug) {
      return $"{Timestamp:yyyy-MM-dd HH:mm:ss} [{Level}-{Caller}-{Source}:{Line}]  {Message})";
    }
    return $"{Timestamp:yyyy-MM-dd HH:mm:ss} [{Level}] {Message}";
  }

  public void ToConsole() {
    Console.ForegroundColor = Level switch {
      LogLevel.Info => ConsoleColor.White,
      LogLevel.Error => ConsoleColor.Red,
      LogLevel.Debug => ConsoleColor.Gray,
      _ => Console.ForegroundColor
    };
    Console.WriteLine(ToString());
    Console.ResetColor();
  }
}

public enum LogLevel {
  Error
  , Info
  , Debug
  , DebugGit
}

File: lib/utils/logger.dart
import 'dart:collection';
import 'dart:io';
import 'package:path/path.dart' as path;
import 'common.dart';
import 'log_message.dart';
import 'log_level.dart';
import 'package:stack_trace/stack_trace.dart' as stack_trace;

class Logger {
  // Properties
  static LogLevel logLevel = LogLevel.info;
  static bool logToConsole = false;
  static bool logToFile = true;
  static final Queue<LogMessage> logMessages = Queue<LogMessage>();

  // Event handling
  static final List<Function()> _onLogMessageAddedListeners = [];
  static void addOnLogMessageAddedListener(Function() listener) {
    _onLogMessageAddedListeners.add(listener);
  }

  static void removeOnLogMessageAddedListener(Function() listener) {
    _onLogMessageAddedListeners.remove(listener);
  }

  // File path
  static String? _logFilePath;
  static Future<String> get logFilePath async {
    _logFilePath ??= path.join(
      await Common.getAppDataPath(),
      'RepoManager.log',
    );
    return _logFilePath!;
  }

  // Public interface
  static void info(
    String message, [
    String caller = '',
    String sourceFilePath = '',
    int lineNumber = 0,
  ]) => log(LogLevel.info, message);

  static void error(
    String message, [
    String caller = '',
    String sourceFilePath = '',
    int lineNumber = 0,
  ]) => log(LogLevel.error, message);

  static void errorWithException(
    String message,
    Exception exception, [
    String caller = '',
    String sourceFilePath = '',
    int lineNumber = 0,
  ]) => log(LogLevel.error, '$message - ${exception.toString()}');

  static void debug(
    String message, [
    String caller = '',
    String sourceFilePath = '',
    int lineNumber = 0,
  ]) => log(LogLevel.debug, message);

  static void debugGit(
    String message, [
    String caller = '',
    String sourceFilePath = '',
    int lineNumber = 0,
  ]) => log(LogLevel.debugGit, message);

  static Future<void> openLogFile() async {
    if (!logToFile) return;

    final logDir = path.dirname(await logFilePath);
    if (!Directory(logDir).existsSync()) {
      await Directory(logDir).create(recursive: true);
    }

    Process.start('open', [await logFilePath]);
  }

  // Implementation
  static void log(LogLevel level, String message) {
    if (level.index > logLevel.index) return;

    final trace = stack_trace.Trace.current();
    final caller = trace.frames[2].member.toString();
    final source = trace.frames[2].uri.toString();
    final line = trace.frames[2].line;

    final logMessage = LogMessage(
      level,
      message,
      caller,
      path.basename(source),
      line ?? -1,
    );
    logMessages.add(logMessage);

    if (logToFile) {
      _logToFile(logMessage);
    }
    if (logToConsole) {
      logMessage.toConsole();
    }

    for (var listener in _onLogMessageAddedListeners) {
      listener();
    }
  }

  static void _logToFile(LogMessage message) async {
    final logDir = path.dirname(await logFilePath);
    if (!Directory(logDir).existsSync()) {
      Directory(logDir).createSync(recursive: true);
    }

    const maxSize = 10 * 1024 * 1024; // 10MB
    final file = File(await logFilePath);
    if (file.existsSync() && file.lengthSync() > maxSize) {
      final backupPath = path.join(logDir, 'RepoManager.log.1');
      if (File(backupPath).existsSync()) {
        File(backupPath).deleteSync();
      }
      file.renameSync(backupPath);
      file.createSync();
    }
    file.writeAsStringSync('$message\n', mode: FileMode.append);
  }
}
